src/cascadia/TerminalControl/TSFInputControl.cpp

```diff
@@ -0,0 +1,380 @@
+    // Return Value:
+    // - <none>
+    void TSFInputControl::_compositionCompletedHandler(CoreTextEditContext sender, CoreTextCompositionCompletedEventArgs const& /*args*/)
+    {
+        // only need to do work if the current buffer has text
+        if (!_inputBuffer.empty())
+        {
+            const auto hstr = to_hstring(_inputBuffer.c_str());
+
+            // call event handler with data handled by parent
+            _compositionCompletedHandlers(hstr);
+
+            // clear the buffer for next round
+            _inputBuffer.clear();
+            _textBlock.Text(L"");
+
+            // tell the input server that we've cleared the buffer
+            CoreTextRange emptyTextRange;
+            emptyTextRange.StartCaretPosition = 0;
+            emptyTextRange.EndCaretPosition = 0;
+
+            // indicate text is now 0
+            _editContext.NotifyTextChanged(emptyTextRange, 0, emptyTextRange);
+            _editContext.NotifySelectionChanged(emptyTextRange);
+
+            // hide the controls until composition starts again
+            _canvas.Visibility(Visibility::Collapsed);
+            _textBlock.Visibility(Visibility::Collapsed);
+        }
+    }
+
+    // Method Description:
+    // - Handler for FocusRemoved event by CoreEditContext responsible
+    //   for removing focus for the TSFInputControl control accordingly
+    //   when focus was forecibly removed from text input control. (TODO GitHub #3644)
+    //   NOTE: Documentation says application should handle this event
+    // Arguments:
+    // - sender: CoreTextEditContext sending the request. Not used in method.
+    // - object: CoreTextCompositionStartedEventArgs. Not used in method.
+    // Return Value:
+    // - <none>
+    void TSFInputControl::_focusRemovedHandler(CoreTextEditContext sender, winrt::Windows::Foundation::IInspectable const& /*object*/)
+    {
+    }
+
+    // Method Description:
+    // - Handler for TextRequested event by CoreEditContext responsible
+    //   for returning the range of text requeted.
+    //   NOTE: Documentation says application should handle this event
+    // Arguments:
+    // - sender: CoreTextEditContext sending the request. Not used in method.
+    // - args: CoreTextTextRequestedEventArgs to be updated with requested range text.
+    // Return Value:
+    // - <none>
+    void TSFInputControl::_textRequestedHandler(CoreTextEditContext sender, CoreTextTextRequestedEventArgs const& args)
+    {
+        // the range the TSF wants to know about
+        const auto range = args.Request().Range();
+
+        try
+        {
+            const auto textRequested = _inputBuffer.substr(range.StartCaretPosition, static_cast<size_t>(range.EndCaretPosition) - static_cast<size_t>(range.StartCaretPosition));
+
+            args.Request().Text(winrt::to_hstring(textRequested.c_str()));
+        }
+        catch (...)
+        {
+            LOG_CAUGHT_EXCEPTION();
+        }

```

Comment:
You can actually `CATCH_LOG()` and replace the entire catch and body :smile:

Reply:
Awesome, didn't exist many months ago ;)